docs/book/developer-guide/advanced-concepts/materializer.md

```diff
@@ -0,0 +1,454 @@
+how this artifact is first serialized and written to the artifact store, and
+then deserialized and read in the next step.
+
+For most data types, ZenML already includes built-in materializers that
+automatically handle artifacts of those data types. For instance, all of the 
+examples from the [Steps and Pipelines](../steps-pipelines/steps-and-pipelines.md)
+section were using built-in materializers under the hood to store and load 
+artifacts correctly.
+
+However, if you want to pass custom objects between pipeline steps, such as a
+PyTorch model that does not inherit from `torch.nn.Module`, then you need to
+define a custom Materializer to tell ZenML how to handle this specific data
+type.
+
+## Building a Custom Materializer
+
+### Base Implementation
+
+Before we dive into how custom materializers can be built, let us briefly
+discuss how materializers in general are implemented. In the following, you
+can see the implementation of the abstract base class `BaseMaterializer`, which
+defines the interface of all materializers:
+
+```python
+from typing import Type, Any
+from zenml.materializers.base_materializer import BaseMaterializerMeta
+
+
+class BaseMaterializer(metaclass=BaseMaterializerMeta):
+    """Base Materializer to realize artifact data."""
+
+    ASSOCIATED_ARTIFACT_TYPES = ()
+    ASSOCIATED_TYPES = ()
+
+    def __init__(self, artifact: "BaseArtifact"):
+        """Initializes a materializer with the given artifact."""
+        self.artifact = artifact
+
+    def handle_input(self, data_type: Type[Any]) -> Any:
+        """Write logic here to handle input of the step function.
+
+        Args:
+            data_type: What type the input should be materialized as.
+        Returns:
+            Any object that is to be passed into the relevant artifact in the
+            step.
+        """
+        # read from self.artifact.uri
+        ...
+
+    def handle_return(self, data: Any) -> None:
+        """Write logic here to handle return of the step function.
+
+        Args:
+            data: Any object that is specified as an input artifact of the step.
+        """
+        # write `data` to self.artifact.uri
+        ...
+```
+
+### Which Data Type to Handle?
+
+Each materializer has an `ASSOCIATED_TYPES` attribute that contains a list of
+data types that this materializer can handle. ZenML uses this information to
+call the right materializer at the right time. I.e., if a ZenML step returns a
+`pd.DataFrame`, ZenML will try to find any materializer that has `pd.DataFrame`
+in its `ASSOCIATED_TYPES`. List the data type of your custom object here to
+link the materializer to that data type.
+
+### What Type of Artifact to Generate
+
+Each materializer also has an `ASSOCIATED_ARTIFACT_TYPES` attribute, which
+defines what types of artifacts are being stored, such as `DataArtifact`,
+`StatisticsArtifact`, `DriftArtifact`, etc. This is not too important for you,
```

Comment:
Is it used in the post-execution workflow? I dont think so. But we can say that potentially it can be used in the post-execution workflow and also in visualizers to visualize artifacts

Reply:
No there is nowhere its used as far as i can see except some visualizers like facets